### DISCUSSION TITLE
Error: refname refs/heads/master not found

### DISCUSSION THREAD

**Discussion by fretagi** [community/community#24104] [New to GitHub]
<p>Hi</p>
<p>I am trying to publish my third site, so I fairly new to github, and I am getting: the following error:</p>
<blockquote>
<p>$ git branch -M main<br>
error: refname refs/heads/master not found<br>
fatal: Branch rename failed</p>
</blockquote>
<p>I have used git-bash, them some code on github account, and on the above stage I got the error.<br>
I am bit stuck, I have search for the error, but I could not fix it.</p>

**Comment by ghost** [comment]
<p>Did you solve the problem ? I take that error too and couldn’t find solution</p>

**Comment by airtower-luna** [comment]
<p>Could the branch possibly be named <code>main</code> already? Github <a href="https://github.com/github/renaming">changed the default name for the default branch</a> from <code>master</code> to <code>main</code> a while ago.</p>

**Comment by fretagi** [comment, marked as answer]
<p>Can I post the link to the solution here? Its an external source… and its in Portuguese:<br>
<a href="https://pt.stackoverflow.com/questions/388005/problema-na-hora-de-fazer-push-no-github" rel="noopener nofollow ugc">https://pt.stackoverflow.com/questions/388005/problema-na-hora-de-fazer-push-no-github</a></p>
<p>Basically what I did was  to run the</p>
<pre><code class="lang-auto">git fetch
git reset --hard origin/master
</code></pre>
<p>them <code>git push</code></p>

**Reply by RitaBallesterosB** [reply]
Gracias, tenía ese mismo error y ejecuté la respuesta en mi máquina virtual y me funcionó

**Reply by RitaBallesterosB** [reply]
ha funcionado para cambiar a la rama main

**Comment by ewink1977** [comment]
<p>I had this error as well, but it cleared after I made my initial commit. That might be the issue?</p>

**Reply by aymankhaled27** [reply]
That was what I missed too, I was searching for solution for over an hour, but at the end I just read your comment "after I made my initial commit" and then i went back to VSCode and found that the commit was not completed .. OMG!
I just completed the commit changes successfully, then linked Netlify to my GitHub repo again, and then the site deployment was completed successfully.

**Reply by melisandoval** [reply]
Had exactly the same problem, and you are right, after the first commit tried again and worked. Thanks for sharing this!

**Reply by arhab213** [reply]
I'm really stupid, like how you can have a master if you have 0 commit, thnx it's really helps my 🫡

**Comment by fretagi** [comment]
<p>I deleted the repo, and started all over again, and it come right. thanks :grinning:</p>

**Comment by damwead** [comment]
<p>same here, it turns out that to make a first push you gotta make a first commit first</p>

**Comment by jeffchavez-dev** [comment]
<p>I also deleted the repo and it worked!</p>

**Comment by neurorepro** [comment]
<p>A branch is a pointer to a commit, so if no commit has been created yet, then no branch exists. So you cannot rename a branch which does not exist. 

The solution is to do a first commit, which will create a branch (master by default), and then you should be able to rename it (now that it exists).</p>

So in other words:
```
git commit -m "Initial commit"
git branch -M main
```

**Reply by HemilTheRebel** [reply]
Thanks a lot for this

**Comment by Fahari** [comment]
<p>Run</p>
<p>git config --global user.email “<a href="mailto:you@example.com">you@example.com</a>”<br>
git config --global user.name “Your Name”</p>
<p>and then make your first commit.</p>
<p>$ git commit -m “first commit”</p>
<p>and now run the command</p>
<p>$ git branch -M main</p>
<p>It worked for me this way.</p>

**Reply by rogermix69** [reply]
obrigado, funcionou pra mim tb

**Comment by AmbassEugene** [comment]
<p>Yeah, this here is the simpler solution.</p>

**Comment by marian123Coman** [comment]
<p>git init<br>
git config --global user.name ‘your name here’<br>
git config --global user.email ‘your email here’<br>
git add .<br>
git commit -m ‘some message’<br>
create new repository in github and copy ssh link<br>
git remote add origin ‘ssh link copied above’<br>
git push -u origin master</p>
<p>That worked for me/</p>

**Comment by paulmelero** [comment]
<p>For me, the problem was that I had untracked files. Once you “git add …” + “git commit …”, you can rename the branch.</p>

**Comment by lunayyko** [comment]
<p>that was why! thanks!</p>

**Comment by leochoo** [comment]
<aside class="onebox allowlistedgeneric" data-onebox-src="https://dev.to/otumianempire/change-branch-name-from-master-to-main-50ei">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/26427048/181082692-ec55c228-0c57-4f24-b977-26a9acca3e20.png" class="site-icon" width="32" height="32">

      <a href="https://dev.to/otumianempire/change-branch-name-from-master-to-main-50ei" target="_blank" rel="noopener nofollow ugc">DEV Community</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/367;"><img src="https://user-images.githubusercontent.com/26427048/181082693-44644644-f1cc-40d0-9cf8-133bfd5feffa.png" class="thumbnail" width="690" height="367" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/5/5/551c9650454aa84969d704c2abc26074dc9dde3c_2_690x367.png, https://aws1.discourse-cdn.com/github/optimized/3X/5/5/551c9650454aa84969d704c2abc26074dc9dde3c_2_1035x550.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/5/5/551c9650454aa84969d704c2abc26074dc9dde3c.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/5/5/551c9650454aa84969d704c2abc26074dc9dde3c_2_10x10.png"></div>

<h3><a href="https://dev.to/otumianempire/change-branch-name-from-master-to-main-50ei" target="_blank" rel="noopener nofollow ugc">Change branch name from master to main</a></h3>

  <p>By google-ing, we'd find a  simple command as git branch -m master main which should rename the maste...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by ferasalsouri** [comment]
<p>You are currently in <em>detached head state</em>. You must checkout a new branch to associate it with the current commit:  <code>git checkout -b new_branch</code></p>

**Comment by dranoer** [comment]
<p>yea it worked for me</p>

**Comment by moyakub** [comment]
<p>Thanks. I had the same issue. Did the first commit. pushed to remote origin first time fine. Thanks again for the tip</p>

**Comment by eguitler** [comment]
<p>It helped me, I needed to make a commit first and after run git branch -M master</p>

**Comment by sprogrammer9** [comment]
<p>i just had to register so i can say this, you saved me a whole lot of stress with this comment… i was using master in my header and i have been stuck at it for hours just for me to stumble upon this comment and i changed “master” to “main” and boom, it worked</p>
<p>thank you buddy…</p>

**Comment by SteelBlueVision** [comment]
If none of the above worked for you, your issue may be the following:

You have a local branch (say `master`) and your local repo knows of a remote branch (say `origin/master`), which may have previously existed, but no longer does (say because the remote repo now uses `origin/main` instead of `origin/master`. When you try to attach your local `master` branch to the remote `origin/master` branch, you are successful (!), because git thinks the remote branch is still out there, even though it no longer is.

Unfortunately this creates a deferred issue for you. As soon as you do a push or pull on `master` you get the dreaded error this thread is about:

```text
Your configuration specifies to merge with the ref 'origin/master'
from the remote, but no such ref was fetched.
```

The solution is simple:

1. Get an up to date snapshot of the remote: `git fetch -p` # The -p is for prune
2. Get rid of your master branch: (`git branch -d master`, when HEAD is positioned on some other branch, say `main`, or some commit, directly by hash):
3. If still present after the fetch/prune (`git branch -r` # Will tell you this info, the `-r` is to limit to remote branches, only), get rid of your local git repo's record of the remote `origin/master` branch (that no longer exists): `git branch -d origin/master`
4. Live happily ever after... (optional)

If this helps you, please upvote so others see this closer to the top, because none of the above solutions will work for this relatively common case that confounds people.

**Reply by airtower-luna** [reply]
An easier way to get rid of no-longer existing remote branches is [`git remote prune`](https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-empruneem).

Note though that occasionally having those around can be useful, specifically if a remote branch has been deleted and you'd still like to have that version. Admittedly that's rare, but I make it a point to warn people about potentially destructive actions. :smile_cat:

**Comment by Allislove** [comment]
In my case, I was trying to create the branch in "an existing repository", but this was not existing, so it followed the github rules, and I had to create a new repository from my project's parent directory.
Note: This will not remove anything from your project, all it will do is create a git repository and you can push it to github without any problems.

```
- echo "# <change-this-by-your-repository-name>" >> README.md
- git init
- git add README.md
- git commit -m "first commit"
- git branch -M home
- git remote add origin https://github.com/<YOUR-USERNAME>/<change-this-by-your-repository-name>.git
- git push -u origin home
```
